File: src/services/assessor.ts
import type { Analysis } from '../models/analysis.js';
import type { Profile } from '../models/profile.js';
import type { Policy } from '../models/policy.js';
import type { RelevanceScore, RiskScore, SuggestedAction } from '../models/adaptation.js';

export interface AssessmentResult {
  relevance: RelevanceScore;
  riskScore: RiskScore;
  strategicValue: string;
  suggestedAction: SuggestedAction;
}

/**
 * Assess the relevance, risk, and suggested action for an analysis
 * given the downstream project's profile and policy.
 */
export function assessRelevance(
  analysis: Analysis,
  profile: Profile,
  policy: Policy,
): AssessmentResult {
  const relevance = computeRelevance(analysis, profile, policy);
  const riskScore = computeRisk(analysis);
  const suggestedAction = computeSuggestedAction(relevance, riskScore);
  const strategicValue = computeStrategicValue(analysis, profile);

  return { relevance, riskScore, strategicValue, suggestedAction };
}

/**
 * Relevance scoring:
 *   - affectedModules overlap with profile.criticalModules → high
 *   - affectedModules overlap with policy.relevantModules  → medium
 *   - affectedModules overlap with policy.ignoredModules   → low
 *   - default: medium
 */
function computeRelevance(
  analysis: Analysis,
  profile: Profile,
  policy: Policy,
): RelevanceScore {
  const modules = analysis.affectedModules;

  // Check critical modules first (highest priority)
  if (
    profile.criticalModules.length > 0 &&
    modules.some((m) => profile.criticalModules.includes(m))
  ) {
    return 'high';
  }

  // Check ignored modules (lowest priority — if everything is ignored, low relevance)
  if (
    policy.ignoredModules.length > 0 &&
    modules.every((m) => policy.ignoredModules.includes(m))
  ) {
    return 'low';
  }

  // Check relevant modules
  if (
    policy.relevantModules.length > 0 &&
    modules.some((m) => policy.relevantModules.includes(m))
  ) {
    return 'medium';
  }

  return 'medium';
}

/**
 * Risk scoring:
 *   - security classification → high
 *   - many files changed (>20) or many additions (>500) → high
 *   - moderate files changed (>5) → medium
 *   - default: low
 */
function computeRisk(analysis: Analysis): RiskScore {
  if (analysis.classification === 'security') {
    return 'high';
  }

  if (analysis.diffStats.filesChanged > 20 || analysis.diffStats.additions > 500) {
    return 'high';
  }

  if (analysis.diffStats.filesChanged > 5) {
    return 'medium';
  }

  return 'low';
}

/**
 * Suggested action based on relevance and risk:
 *   - high relevance + low risk  → adopt
 *   - high relevance + high risk → adapt-partially
 *   - medium relevance           → monitor
 *   - low relevance              → ignore
 */
function computeSuggestedAction(
  relevance: RelevanceScore,
  risk: RiskScore,
): SuggestedAction {
  if (relevance === 'high' && risk === 'low') {
    return 'adopt';
  }
  if (relevance === 'high' && risk === 'high') {
    return 'adapt-partially';
  }
  if (relevance === 'high') {
    // high relevance + medium risk
    return 'adopt';
  }
  if (relevance === 'medium') {
    return 'monitor';
  }
  return 'ignore';
}

/**
 * Compute strategic value from the overlap between analysis intent
 * and the profile's priorities.
 */
function computeStrategicValue(analysis: Analysis, profile: Profile): string {
  if (profile.priorities.length === 0) {
    return `Relevant ${analysis.classification} change in ${analysis.affectedModules.join(', ') || 'unknown modules'}.`;
  }

  const intentLower = analysis.intent.toLowerCase();
  const matchingPriorities = profile.priorities.filter((p) =>
    intentLower.includes(p.toLowerCase()),
  );

  if (matchingPriorities.length > 0) {
    return `Aligns with priorities: ${matchingPriorities.join(', ')}.`;
  }

  return `${analysis.classification.charAt(0).toUpperCase() + analysis.classification.slice(1)} change; review for alignment with project goals.`;
}

File: code_adapt/services/assessor.py
"""Relevance and risk assessment logic."""

from __future__ import annotations

from dataclasses import dataclass

from ..models import (
    Analysis,
    Policy,
    Profile,
    RelevanceScore,
    RiskScore,
    SuggestedAction,
)


@dataclass
class AssessmentResult:
    relevance: RelevanceScore
    risk_score: RiskScore
    strategic_value: str
    suggested_action: SuggestedAction


def assess_relevance(analysis: Analysis, profile: Profile, policy: Policy) -> AssessmentResult:
    relevance = _compute_relevance(analysis, profile, policy)
    risk = _compute_risk(analysis)
    action = _compute_suggested_action(relevance, risk)
    value = _compute_strategic_value(analysis, profile)
    return AssessmentResult(
        relevance=relevance, risk_score=risk, strategic_value=value, suggested_action=action
    )


def _compute_relevance(
    analysis: Analysis, profile: Profile, policy: Policy
) -> RelevanceScore:
    modules = analysis.affected_modules
    if profile.critical_modules and any(m in profile.critical_modules for m in modules):
        return RelevanceScore.HIGH
    if policy.ignored_modules and all(m in policy.ignored_modules for m in modules):
        return RelevanceScore.LOW
    if policy.relevant_modules and any(m in policy.relevant_modules for m in modules):
        return RelevanceScore.MEDIUM
    return RelevanceScore.MEDIUM


def _compute_risk(analysis: Analysis) -> RiskScore:
    if analysis.classification.value == "security":
        return RiskScore.HIGH
    if analysis.diff_stats.files_changed > 20 or analysis.diff_stats.additions > 500:
        return RiskScore.HIGH
    if analysis.diff_stats.files_changed > 5:
        return RiskScore.MEDIUM
    return RiskScore.LOW


def _compute_suggested_action(
    relevance: RelevanceScore, risk: RiskScore
) -> SuggestedAction:
    if relevance == RelevanceScore.HIGH and risk == RiskScore.LOW:
        return SuggestedAction.ADOPT
    if relevance == RelevanceScore.HIGH and risk == RiskScore.HIGH:
        return SuggestedAction.ADAPT_PARTIALLY
    if relevance == RelevanceScore.HIGH:
        return SuggestedAction.ADOPT
    if relevance == RelevanceScore.MEDIUM:
        return SuggestedAction.MONITOR
    return SuggestedAction.IGNORE


def _compute_strategic_value(analysis: Analysis, profile: Profile) -> str:
    if not profile.priorities:
        modules_str = ", ".join(analysis.affected_modules) or "unknown modules"
        return f"Relevant {analysis.classification.value} change in {modules_str}."

    intent_lower = analysis.intent.lower()
    matching = [p for p in profile.priorities if p.lower() in intent_lower]
    if matching:
        return f"Aligns with priorities: {', '.join(matching)}."

    label = analysis.classification.value.capitalize()
    return f"{label} change; review for alignment with project goals."
